### PR TITLE
KAFKA-15302: Stale value returned when using store.all() with key deletion [docs]

### PR DESCRIPTION
**Pull request by jinyongchoi**
Related Issue
[KAFKA-15302](https://issues.apache.org/jira/browse/KAFKA-15302)

This PR adds documentation regarding the potential return of stale values when using 'withCachingEnabled' and deleting keys during store iteration.
Since there's a current solution involving flush(), modifications may not be made immediately. Hence, it's necessary to explicitly state in the documentation. 

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: docs/streams/developer-guide/memory-mgmt.html [hunk 1]
```diff
@@ -151,6 +151,10 @@
     Serdes.Long())
   .withCachingEnabled();</code></pre>
       <p>Record caches are not supported for <a class="reference internal" href="processor-api.html#streams-developer-guide-state-store-versioned"><span class="std std-ref">versioned state stores</span></a>.</p>
+      <p><strong>Caution:</strong> When using <code class="docutils literal"><span class="pre">withCachingEnabled()</span></code>,
+        if you <code class="docutils literal"><span class="pre">delete()</span></code> a key from the Store while iterating(e.g., <code class="docutils literal"><span class="pre">KeyValueIterator</span></code>) through the keys,
```
Comment (mjsax) [inline docs/streams/developer-guide/memory-mgmt.html]:
```suggestion
        if you <code class="docutils literal"><span class="pre">delete()</span></code> a key from the store while iterating (e.g., <code class="docutils literal"><span class="pre">KeyValueIterator</span></code>) through the keys,
```

Comment (mjsax) [reply docs/streams/developer-guide/memory-mgmt.html]: Cannot remember from the ticket, but was this limited to deletes, or also updates?

File: docs/streams/developer-guide/memory-mgmt.html [hunk 2]
```diff
@@ -151,6 +151,10 @@
...
+        the value of the key that hasn't been flushed from the cache may be returned as a stale value.
+        Therefore, when deleting, you must <code class="docutils literal"><span class="pre">flush()</span></code> the cache before the iterator.</p>
```
Comment (mjsax) [inline docs/streams/developer-guide/memory-mgmt.html]: I would rephrase this sentenc.
```
To avoid reading stale data, you can `flush()` the store before creating the iterator. Note, that flushing too often can lead to performance degration if RocksDB is used, so we advice to avoid flushing manually in general.

**Review by mjsax — COMMENTED** [review]
Thanks for the PR. Couple of minor suggestions.

File: docs/streams/developer-guide/memory-mgmt.html (hunk 1 above)
Comment (jinyongchoi) [reply docs/streams/developer-guide/memory-mgmt.html]: > Cannot remember from the ticket, but was this limited to deletes, or also updates?

Yes, that's correct.
This affects the snapshot created in store.all(), causing it to read the values from rocksdb before the cache is flushed.

File: docs/streams/developer-guide/memory-mgmt.html (hunk 2 above)
Comment (jinyongchoi) [reply docs/streams/developer-guide/memory-mgmt.html]: > I would rephrase this sentenc.
> ```
> To avoid reading stale data, you can `flush()` the store before creating the iterator. Note, that flushing too often can lead to performance degration if RocksDB is used, so we advice to avoid flushing manually in general.

Ok. I will change.

Comment (jinyongchoi) [reply docs/streams/developer-guide/memory-mgmt.html]: @mjsax Could you verify the changes I fixed?

memory-mgmt html

**Comment by mjsax** [comment]
Thanks for the PR @jinyongchoi -- and sorry for the delay -- seems the PR slipped through the cracks...

Merged to `trunk` and cherry-picked to `3.8` branch (just in time for 3.8 release :))
